KOGITO-3527 Work Item Definition Parser

hasys opened the pull request:
**JIRA**: [KOGITO-3527](https://issues.redhat.com/browse/KOGITO-3527)

**Business Central**: not affected

**VS Code**: [here](https://drive.google.com/file/d/1mXCEVt4qSbagu81t80u0P3hehnbK8DQa/view?usp=sharing)

**Standalone Editor**: [here](https://drive.google.com/file/d/17Ge1YgAhCPubOqR4f34-7YVIUq1NgLM-/view?usp=sharing)

Standalone Editor can be tested using https://github.com/bpmn-tutorials/EmbeddedBpmnEditor

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

* Retest PR: <b>jenkins retest this</b>
* A full downstream build: <b>jenkins do fdb</b>
* A compile downstream build: <b>jenkins do cdb</b>
* A full production downstream build: <b>jenkins do product fdb</b>
* An upstream build: <b>jenkins do upstream</b>
</details>

hasys commented:
jenkins do cdb

domhanak commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/main/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientService.java:
```diff
@@ -48,6 +50,7 @@
 public class WorkItemDefinitionStandaloneClientService implements WorkItemDefinitionClientService {
 
     private static final String RESOURCE_ALL_WID_PATTERN = "*.wid";
+    private static final String RESOURCE_GLOBAL_DIRECTORY_WID_PATTERN = "global/*.wid";
```
@romartin @hasys @ederign As tested, this only allows users to work with one project per workspace in VSCode:
Example:
1. Open Folder `RootFolder` in VSCode
2. Generate you businessapp in this folder `mybusinessapp` (with global, src and pom.xml)
3. Create some process in resources (org/kie/businessapp) and create a WorkItems.wid with Email definition.
4. Open the process and see Email wid in palette, Close the editor.
5. Add a new WID definition file into `global` folder

Actual results:
WIDs in global won't be found, because ResourceContentService from kogito searches from root of openned folder, so `RootFolder/mybusinessapp/global` won't be matched.

Expected result:
Here I am not sure what to expect I see two option we can pursue:
1. Change to `**/global/*.wid` to allow users to work with multiple projects in one workspace (add unnecesarry WIDs to different project)
2. Document this and direct users to use one VSCode instance per project?

WDYT?

hasys replied:
Hi @domhanak,

good point. I think this behavior is expected, see [Kogito Tooling](https://github.com/kiegroup/kogito-tooling/blob/2a5ab955f9f61ffb6ed41e657da87adaf7d6cee8/packages/vscode-extension/src/VsCodeResourceContentService.ts#L42), so option 2 `Document this and direct users to use have their project root openned in VSCode as root.` makes more sense to me.

Also it can be an option 3:
3. Kogito Tooling will provide filter for project isolated global directory so we can just call a filter as we are doing for same directory as BPMN file.

Something like:

```java
resourceContentService.list(RESOURCE_ALL_WID_PATTERN, ResourceListOptions.globalFolder())
```

instead of:
```java
resourceContentService.list(RESOURCE_ALL_WID_PATTERN, ResourceListOptions.assetFolder())
```

(see **ResourceListOptions**).

I know it is not supposed to have _global_ folder in Kogito in long term and it is there only for Business Central compatibility, so we can define it as `@Deprecated` and document why is so right from the start.

domhanak commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-kogito-runtime/src/test/java/org/kie/workbench/common/stunner/kogito/client/services/WorkItemDefinitionStandaloneClientServiceTest.java:
```diff
@@ -132,7 +132,7 @@ public void testNoWorkItemsLoadPreset() {
                                                                new WorkItemIconCacheImpl(new BPMNStaticResourceContentService(promises)));
         call();
-        assertEquals(1, registry.items().size());
+        assertEquals(0, registry.items().size());
```
@hasys As discussed off PR, let's add some case where the service returns some relevant items and tests that stunner deals with them properly. Could be the one we discovered, where there is no global directory in returned list.

domhanak reviewed (changes requested):
Tested general use cases in VSCode with our `kie-kjar-archetype`
I generated the project structure and openned folder `mybusinessap` in VSCode.

- Users can add  `*.wid` files into `/global` directory and BPMN editor detects these. :heavy_check_mark: 

- Users can add  `*.wid` files into same directory as BPMN Process and BPMN editor detects these. :heavy_check_mark: 

- WID parser shouldn’t fail on unknown/unsupported/engine only parameters see KOGITO-4372 for more details. :heavy_check_mark: 

- All supported by engine MVEL types should be converted to Java types during WID loading. :heavy_check_mark: 

- It should be possible to add an icon path inside of the global directory and set up an icon in base 64 format. :x:
- Custom Icons can be set by placing them in `global` directory :x: (Does not work, icon in general would not change)

Icons need to be fixed, also please check https://github.com/kiegroup/kie-wb-common/pull/3552#discussion_r580247776 for issue with multiple projects in one workspace.

domhanak commented:
Also, just to raise awareness. Issue: https://issues.redhat.com/browse/KOGITO-3528 is often encountered during testing of this PR, as there is no guarantee that two collaborating users have the same environment (same WID's in their projects). Workaround is to use a versioning system, but we should consider this in higher level applications (e.g. sharing of process mockups etc.)

hasys commented:
Hi @domhanak,

great job!

>     * It should be possible to add an icon path inside of the global directory and set up an icon in base 64 format. ❌
> 
>     * Custom Icons can be set by placing them in `global` directory ❌ (Does not work, icon in general would not change)
 
Base 64 works for you, right? So I need to checkout only separate files, am I correct?

Thank you @domhanak will take a look at found issues.

hasys commented:
jenkins do cdb

romartin reviewed (approved):
@hasys everything looks great to me! Great quality of code as well. Thanks, great worrk! :+1:

hasys commented:
jenkins do cdb

hasys commented:
jenkins do cdb

hasys commented:
Hi @romartin, @domhanak,

I fixed issue with icons and added more tests as discussed. VS Code extension and standalone editor are updated and ready for test. https://github.com/bpmn-tutorials/EmbeddedBpmnEditor repository also updated to the latest version, so you can use it if needed.

Thank you!

domhanak commented:
As usual, I generated the project structure and openned folder mybusinessap in VSCode.

*    Users can add *.wid files into /global directory and BPMN editor detects these. :heavy_check_mark:

*   Users can add *.wid files into same directory as BPMN Process and BPMN editor detects these. :heavy_check_mark:

*   WID files located outside global directory and outside of directory of BPMN file are not shown in palette. :heavy_check_mark: 

*  WID parser shouldn’t fail on unknown/unsupported/engine only parameters see KOGITO-4372 for more details. :heavy_check_mark:

*  All supported by engine MVEL types should be converted to Java types during WID loading. heavy_check_mark

*  It should be possible to add an icon path inside of the global directory and set up an icon in base 64 format.  :heavy_check_mark: 

* it is possible to setup an icon in base64 format :heavy_check_mark: 

* Custom Icons can be set by placing them in global directory :heavy_check_mark: 

This PR is good to merge.

@romartin @hasys LGTM, hopefully I didn't miss anything, above scenarios will be covered in E2E tests in VSCode to possible extent.
I reported a new issue https://issues.redhat.com/browse/KOGITO-4648 that closely relates to the blocker https://issues.redhat.com/browse/KOGITO-3528

romartin commented:
thanks @domhanak @hasys !! Merging!
